### PR TITLE
RFC: TensorFlow Canonical Type System

### PR DESCRIPTION
**Pull request by mdanatg**
# TensorFlow Canonical Type System

| Status        | Draft       |
:-------------- |:---------------------------------------------------- |
| **RFC #**     | [https://github.com/tensorflow/community/pull/208](https://github.com/tensorflow/community/pull/https://github.com/tensorflow/community/pull/208) |
| **Author(s)** | Dan Moldovan (mdan@google.com) |
| **Sponsor**   | Gaurav Jain (gjn@google.com)                 |
| **Updated**   | 2020-02-19                                           |

## Objective

This RFC proposes a new TensorFlow module and namespace (`tf.types`) dedicated to storing implementation-free type definitions, similar to C++ header files. This module has no other dependencies inside TensorFlow, so any other internal module can depend on it to ensure interoperability without the risk of creating circular dependencies. These definitions can also be used by external users, for example in pytype annotations.
The RFC focuses on the Python API, however the design should be reviewed with cross-language consistency in mind.

### CODE REVIEW

**Comment by gabrieldemarmiesse** [comment]
Very nice proposal! In Tensorflow addons, we've typed the public API, and without this RFC, this is quite a challenge. We've had custom types that we would be happy to get rid of. See https://github.com/tensorflow/addons/blob/master/tensorflow_addons/utils/types.py

Could we also have types which correspond to all valid input for common tensorflow functions? For example, we have thinks like `TensorLike` which refers to anything that can be converted to a tensor. And we use this type a lot in the public API. It it could be taken care of in the TF side, it would be nice.

A good way to ensure that users find all the types they need is to add type hintd to all the public functions and methods in TF, but it's quite some work :)

**Comment by mdanatg** [comment]
@gabrieldemarmiesse Yes, `TensorLike` sounds like a broadly useful definition - I'll add it to the list. The aim is indeed to use these type hints in all public declarations, over time.

File: rfcs/20200211-tf-types.md [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+   
+ * Low-level execution primitives
+   * `Graph` - GraphDef/IR program
+     * `FunctionGraph` - IR of a single concrete function
+
+### Alternatives Considered
+* N/A
+
+### Performance Implications
+* No performance implications expected. At most, we are adding a small number of levels to the class tree of some objects.
```
Comment (superbobry) [inline rfcs/20200211-tf-types.md]: One caveat here is that `isinstance` checks involving ABCs are significantly slower than that with non-abstract classes. This is because `abc.ABCMeta` [implements](https://github.com/python/cpython/blob/1246d892038a693304549f8574e6c2784b91589a/Lib/abc.py#L96) `__instancecheck__` so `isinstance` has to jump back to pure Python instead of doing MRO traversal in C:

```python
>>> import abc                                                                                                                                                                                
>>> class A(abc.ABC): pass                                                                                                                                                                                            
>>> class B: pass                                                                                                                                                                             
>>> class SA(A): pass                                                                                                                                                                         
>>> class SB(B): pass                                                                                                                                                                         
>>> sa = SA()                                                                                                                                                                                 
>>> sb = SB()                                                                                                                                                                                 
>>> %timeit isinstance(sa, A)                                                                                                                                                                 
237 ns ± 0.45 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
>>> %timeit isinstance(sb, B)                                                                                                                                                                 
51.2 ns ± 0.245 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
```

While this is of course an artificial benchmark, and a single `isintance` check is still "just" ~200ns (on my fairly slow laptop), TensorFlow does a lot of such checks, and the slowdown might as well be noticeable in more realistic workloads.

File: rfcs/20200211-tf-types.md [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+### Initial Type Hierarchy
+TensorFlow generally adopts an incremental development method. This RFC aims to remain consistent with that.
+
+Below are listed the major types presently used in TensorFlow. All types included in this list are subject to [normal compatibility rules](https://www.tensorflow.org/guide/versions), so they are unlikely to change in the future. It is therefore preferable to maintain a strict minimum of orthogonal declarations and carefully vet any additions.
+
+Most of these symbols will not be initially exported as public symbols. Only internal submodules will be able to use unexported types. The unexported types may be gradually exposed under `tf.types` or under `tf.types.experimental`.
+
+The initial type hierarchy is focused on V2 symbols. We expect to encounter places where these symbols would not be compatible with V1 code; in such cases, the V1 symbols will not be affected.
+
+#### Types created by this RFC
```
Comment (superbobry) [inline rfcs/20200211-tf-types.md]: Are the specifics of each type out of scope for this RFC? i.e. which methods will be available on a `Shape` or `Tensor`?

File: rfcs/20200211-tf-types.md [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+
+These types are raised for discussion by this RFC, but are not part of the original implementation, unless they are strictly required for consistency (to be determined during the initial submission).
+
+Many of these are expected to be required when breaking the cyclic dependencies that currently exist between submodules. However, it is hoped that opening them up for discussion early can help create a more coherent type system.
+
+* Container types
+
+  * `Composite` - low-level static structure (opaque to GraphDef/IR)
+  * `Module` - builder for structures of `Variables` (invisible to GraphDef/IR)
+  * `Optional` - basic programming construct
```
Comment (superbobry) [inline rfcs/20200211-tf-types.md]: Is this an alias to `typing.Optional` or a separate type? If the latter, could you give an example when one might need a `tf.types.Optional` and not `typing.Optional`?

File: rfcs/20200211-tf-types.md [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+* Core tensor types
+
+  * `DType`
+  * `Shape`
+  * `Tensor` - generic dense tensor
+
+    * `Symbol` - the regular graph tensor
+    * `Value` - eager tensors
+
+  * `TensorLike` - any type that can be implicitly converted to `Tensor` (see for example https://github.com/tensorflow/addons/blob/master/tensorflow_addons/utils/types.py)
```
Comment (superbobry) [inline rfcs/20200211-tf-types.md]: How will `TensorLike` interact with custom ->`Tensor` conversion machinery? If a user registers a new conversion function, should they also [register](https://docs.python.org/3/library/abc.html#abc.ABCMeta.register) a "virtual" subclass of `TensorLike`?

File: rfcs/20200211-tf-types.md [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+The type definitions are consistent with `isinstance` and `issubclass`. For example, `isinstance(tf.Tensor, tf.types.Tensor) == True`.
+
+### General Principles
+This module should not contain any implementation code. An advantage of that is that users exploring the implementation of specific types will not need to inspect this module. However, users who do not wish to inspect the code may visit the documentation of these generic types to better understand specifically what are the concrete subclasses of this type expected to do.
+
+The `tf.types` module may depend on external packages (such as `numpy`) _strictly for the purpose of defining type annotations and documentation_. No dependencies to other TensorFlow interfaces are allowed. Any dependencies on external packages which themselves depend on TensorFlow are expressly forbidden.
+
+Changes definitions inside `tf.types` must be approved by TensorFlow leads, and typically should be accompanied by an RFC.
+
+All type declarations are compatible with [pytype](https://github.com/google/pytype).
```
Comment (superbobry) [inline rfcs/20200211-tf-types.md]: Do you expect them to also be compatible with other type checkers, e.g. [`mypy`](http://mypy-lang.org) or [`pyre`](https://pyre-check.org/)?

File: rfcs/20200211-tf-types.md (hunk 1 above)
Comment (mdanatg) [reply rfcs/20200211-tf-types.md]: Very interesting. It sounds like we may want to avoid adding ABCMeta superclasses in instances when performance is critical, like eager tensors.

File: rfcs/20200211-tf-types.md (hunk 2 above)
Comment (mdanatg) [reply rfcs/20200211-tf-types.md]: I was hoping so, although I think we'll need to define them sooner than later if type annotations are to be useful.

File: rfcs/20200211-tf-types.md (hunk 3 above)
Comment (mdanatg) [reply rfcs/20200211-tf-types.md]: I'll add a clarification - `tf.types.Optional` is an alias to `tf.data.experimental.Optional`, so a separate type from `typing.Optional`. I think the main differences are that it cannot be `None` and exposes `has_value()` `get_value()`.

File: rfcs/20200211-tf-types.md (hunk 4 above)
Comment (mdanatg) [reply rfcs/20200211-tf-types.md]: Good question. Since using ABC may have performance implications as you pointed out, we may need to think of a different mechanism, like a concrete `TensorLike` superclass (instead of a `Union` generic) that also defines `to_tensor` or defining a dunder method similar to `__ndarray__`.

File: rfcs/20200211-tf-types.md (hunk 5 above)
Comment (mdanatg) [reply rfcs/20200211-tf-types.md]: Updated the text - the hope is that so long as we rely on `typing`, static definitions and compliance with the relevant PEPs, they should be compatible with most (all?) static type checkers.

File: rfcs/20200211-tf-types.md (hunk 1 above)
Comment (superbobry) [reply rfcs/20200211-tf-types.md]: One option is to define these types without `abc.ABCMeta` as a metaclass, i.e. make them abstract by convention. The way `abc.ABCMeta` enforces an implementation (at instantiation time instead of definition time) is likely not relevant/useful for `tf.types`. Note however, that this also makes these type non-extensible which could be an issue for `TensorLike`.

File: rfcs/20200211-tf-types.md [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+ * Distributed types
+   * `DistributedDataset` - collective ETL
+   * `DistributedIterator` - collective iterator
+   
+ * Low-level execution primitives
+   * `Graph` - GraphDef/IR program
+     * `FunctionGraph` - IR of a single concrete function
+
+### Alternatives Considered
+* N/A
```
Comment (gabrieldemarmiesse) [inline rfcs/20200211-tf-types.md]: We should consider the alternative where not everything is in the same module and we don't use the trick of making this:

> variables are instances of `tf.Variable`, which is now a subclass of `tf.types.Variable`

I like the idea of having types in tensorflow, but this whole ABC black magic of avoiding dependecies is done because internally, Tensorflow has many, many circular dependencies. In python, the more black magic we do, the more it will come back to bite us, see https://github.com/tensorflow/community/pull/208/files#r382246511  . 

Python types are not supposed to be used in this way and I'm afraid that we're going to have many problems in the future just for the sake of supporting circular dependencies in the TF codebase. Like for example some type system being confused about our types, or we'll realize the need to redeclare all the classes' interfaces in tf.types, breaking the DRY rule.

The alternative is to declare types like `AnyTensor = Union[tf.Tensor, tf.sparse.SparseTensor]` and to fix tensorflow circular dependencies internally. Types can be declared in their corresponding files (ex: `AnyTensor` is defined in the file defining the `Tensor` class).

In short, making Tensorflow types without any ABC tricks shouldn't create new circular dependencies, they'll just make the existing ones come to light.

File: rfcs/20200211-tf-types.md (hunk 4 above)
Comment (superbobry) [reply rfcs/20200211-tf-types.md]: I think that regardless of the implementation, the relationship between `TensorLike`, `tf.convert_to_tensor` and `tf.register_tensor_conversion_function` should be specified to avoid confusion. 

Personally, I'm fine with `TensorLike` being non-extensible, i.e. if you want your custom type to be `TensorLike`, convert it to such explicitly.

Comment (tomhennigan) [reply rfcs/20200211-tf-types.md]: My initial expectation is that `isinstance(x, tf.types.TensorLike)` would work for anything registered with `tf.register_tensor_conversion_function`. I think we cannot do that while requiring `TensorLike` to be a protocol (e.g. np.ndarray doesn't define to_tensor, but can be converted to one).

File: rfcs/20200211-tf-types.md [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+Developers of modules internal to TensorFlow can use this module to avoid creating circular dependencies.
+
+## Design Proposal
+
+### The `tf.types` Namespace / Module
+All the declarations exposed under the `tf.types` namespace reside in the `python/types/*.py` module. These are [abstract base classes](https://docs.python.org/3.7/library/abc.html) with a bare minimum of method definitions and minimal or no implementation, which serve to formalize and document the contract of common types such as `Tensor`, `Variable`, etc.
+
+These definitions may be used as PEP 484 type hints, although in some cases they may be type- or shape- erased (for example, `tf.types.Tensor` may not necessarily be parametrized by `dtype` or `shape`). Note however that designs which parametrize on shape do exist, see for instance [tensorflow#31579](https://github.com/tensorflow/tensorflow/issues/31579).
+
+The type definitions are consistent with `isinstance` and `issubclass`. For example, `isinstance(tf.Tensor, tf.types.Tensor) == True`.
```
Comment (gabrieldemarmiesse) [inline rfcs/20200211-tf-types.md]: I believe this should not be a requirement as this is not a requirements for python types to support `isinstance`. It also restricts us: we can't use anything in the `typing` module and protocols. Ex:

```
from typing import List
import tensorflow as tf

TensorLike = Union[tf.Tensor, List[float]]
isinstance([3.8, 23.4], TensorLike)
```

```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-10-0ff96ca65533> in <module>
----> 1 isinstance([3.8, 23.4], TensorLike)

~/softwares/python/anaconda/lib/python3.7/typing.py in __instancecheck__(self, obj)
    709
    710     def __instancecheck__(self, obj):
--> 711         return self.__subclasscheck__(type(obj))
    712
    713     def __subclasscheck__(self, cls):

~/softwares/python/anaconda/lib/python3.7/typing.py in __subclasscheck__(self, cls)
    717             if cls._special:
    718                 return issubclass(cls.__origin__, self.__origin__)
--> 719         raise TypeError("Subscripted generics cannot be used with"
    720                         " class and instance checks")
    721

TypeError: Subscripted generics cannot be used with class and instance checks
```

I believe this RFC should include some examples of implementation to make things clearer.

File: rfcs/20200211-tf-types.md (hunk 1 above)
Comment (mdanatg) [reply rfcs/20200211-tf-types.md]: Agreed. Just to be clear, the types would be non-extensible using ABCs registration mechanism, and one could still extend using traditional subclassing.
I added a preliminary note - PTAL.

File: rfcs/20200211-tf-types.md (hunk 6 above)
Comment (mdanatg) [reply rfcs/20200211-tf-types.md]: I think I agree with the broader comment: we want to avoid, rather than facilitate, circular dependencies. This RFC specifically aims to support the extract interface pattern for removing such circular dependencies.

That said, I fear the pattern is not a universal solution. For example, in the Keras instance you mentioned, the circular dependency cannot be broken by extracting types and requires moving some code around.

I'm not sure I fully follow the `Variable` example though, could you elaborate on it? The main idea is to have the types defined in a module separated from all others, so for example if you want to recognize a Variable you can import just `tf.types`, rather than the entire `tf` module.

By "ABC tricks" I suspect you are referring to `ABCMeta.register`? There are no plans currently to use it (in fact we are considering not using `abc` at all, see one of the comments above); I hope it will not be needed.

File: rfcs/20200211-tf-types.md (hunk 7 above)
Comment (mdanatg) [reply rfcs/20200211-tf-types.md]: Interesting. It seems like we need to think through the interaction between classical interfaces, protocols and how they play with injection mechanisms like `register_tensor_conversion_function`. Ideally we wouldn't need so many mechanisms for the same thing.

Regarding protocols, I suspect [@runtime_checkable](https://www.python.org/dev/peps/pep-0544/#runtime-checkable-decorator-and-narrowing-types-by-isinstance) can help. Although, it could come with a performance hit, as in the case of ABCMeta.

I'll look into elaborating this.

File: rfcs/20200211-tf-types.md (hunk 6 above)
Comment (gabrieldemarmiesse) [reply rfcs/20200211-tf-types.md]: Here I'm referring to the trick of having all public classes of tensorflow subclass a class defined in tf.types, I think it's a brittle trick to enable the use of `isinstance` and avoid using the `typing` module.

>  if you want to recognize a Variable you can import just tf.types, rather than the entire tf module.

I see, but what is the benefit of that? As a user, what bonus do I get by doing "from tensorflow.types import Variable" rather than `from tensorflow import Variable`? I'm sure there is a reason, but it's not clear in this RFC and we would benefit from a small paragraph detailing the benefits.

A side note for the user experience: A user will first try to use `tf.Variable` as a type hint before reading the docs when he/she needs it. It's not very intuitive to put it in a separate module.

File: rfcs/20200211-tf-types.md (hunk 7 above)
Comment (gabrieldemarmiesse) [reply rfcs/20200211-tf-types.md]: Thanks! :)

File: rfcs/20200211-tf-types.md (hunk 6 above)
Comment (mdanatg) [reply rfcs/20200211-tf-types.md]: I added this as a discussion topic. I'm inclined to agree - we don't need both `tf.types.Variable` and `tf.Variable`, even if it was safe to use either. In that case, `tf.types` would only have entries for types which do not already belong elsewhere.

The Python [types](https://docs.python.org/3.8/library/types.html) module uses a similar pattern.

Related, if you have additional thoughts on how `typing` could enable better decoupling, both internally and externally, please do suggest them.

File: rfcs/20200211-tf-types.md (hunk 4 above)
Comment (mdanatg) [reply rfcs/20200211-tf-types.md]: Added a section about this - PTAL.

It turns out that protocols do support static type checking, and can support isinstance in newer Python versions too.

I'm not sure that `register_tensor_conversion_function` can be made compatible with static type checking. `ABCMeta.register` might be the equivalent of it, but it lacks the means to specify the conversion logic if you don't control the class (e.g. how do you specify the conversion logic for  `list` with ABCMeta.register?). And besides, if we replace `register_tensor_conversion_function`, we might as well opt for the leaner protocols (`hasattr` seems slightly faster than a bare `isinstance`). So my preference would be to use protocol as alternative to  `register_tensor_conversion_function`, and eventually deprecate it.

File: rfcs/20200211-tf-types.md (hunk 7 above)
Comment (mdanatg) [reply rfcs/20200211-tf-types.md]: I added a section proposing the use of protocols for custom Tensor types - I think that is in line with the use of `typing` that you suggested as well.

Comment (gabrieldemarmiesse) [reply rfcs/20200211-tf-types.md]: Thank you very much, I'll take a look whenever I can.

File: rfcs/20200211-tf-types.md [hunk 8]
```diff
@@ -0,0 +1,230 @@
...
+* New minor version. Existing classes (`tf.Tensor`) will become subclasses of the new type interfaces.
+* Most subcomponents of TF (Lite, distributed, function, SavedModel) will depend on this new module, although their functionality is not impacted.
+* Libraries which depend on TensorFlow are encouraged to refer to `tf.types` definitions, rather than the concrete implementations for better future compatibility.
+
+### User Impact
+* Users will see a new `tf.types` module, that may be referenced from documentation and type annotations.
+
+
+## Questions and Discussion Topics
+
```
Comment (qlzh727) [inline rfcs/20200211-tf-types.md]: How would the new type system unified with https://www.tensorflow.org/api_docs/python/tf/TypeSpec?

I think having both name space might confuse user about which to use, and we probably want to unify them into one if possible.

Comment (mdanatg) [reply rfcs/20200211-tf-types.md]: Good question, and I agree it would be great to either unify them, or think of ways to make the differences clearer and more intuitive - perhaps we should begin to refer to them as [TF] Python types and TF [native] types.
The distinctions blur when using type annotations (e.g. if a tf.function had both an input signature and type annotations, we need to at least make sure they're consistent).
The most intuitive path toward unification ought be to use Python type annotations to specify a `TypeSpec`, although the type annotation system is not yet powerful enough.

File: rfcs/20200211-tf-types.md [hunk 9]
```diff
@@ -0,0 +1,293 @@
...
+...   ...
+```
+
+In such cases, `tf.function` is expected to verify that the type annotation matches the `input_signature`.
+
+In the long term, this RFC recommends that type annotations fully replace the `input_signature`, so far as the Python type annotation system allows it:
+
+```
+>>> @function
+... def f(x: Tensor[int32, [3]]):
```
Comment (alextp) [inline rfcs/20200211-tf-types.md]: How can we allow Tensor[int32, [3]]? Does this mean Tensor have to have a `__getitem__` classmethod?

Comment (mdanatg) [reply rfcs/20200211-tf-types.md]: That example was incorrect, sorry about that. Updated to a tested version.

File: rfcs/20200211-tf-types.md [hunk 10]
```diff
@@ -0,0 +1,295 @@
...
+
+In such cases, `tf.function` is expected to verify that the type annotation matches the `input_signature`.
+
+In the long term, this RFC recommends that type annotations fully replace the `input_signature`, so far as the Python type annotation system allows it:
+
+```
+>>> class BatchSize(Dimension):
+...   value = 32
+>>> @function
+... def f(x: Tensor[Int32, Shape1D[BatchSize]]):
```
Comment (gabrieldemarmiesse) [inline rfcs/20200211-tf-types.md]: What would be the recommended way to  type functions which accept multiple dtype? Would Union work? For example: `def f(x: Tensor[Union[tf.float32, tf.float64], Shape[32]]`.

File: rfcs/20200211-tf-types.md [hunk 11]
```diff
@@ -0,0 +1,295 @@
...
+...   ...
```
Comment (gabrieldemarmiesse) [inline rfcs/20200211-tf-types.md]: Why Shape1D rather than Shape? Could we also just use a tuple to specify the shape? It would lead to clearer signatures.

Comment (mdanatg) [reply rfcs/20200211-tf-types.md]: I updated the RFC test to avoid prescribing a certain scheme for type annotations, and use whatever works best when we move to implementing it. It would be great to be able to use a tuple and we should do it if it's possible. In the few tests that I'm aware of [1], certain limitations in the type annotation system precluded the use of tuples, and forced using different type for different tensor ranks (e.g. Tensor1D, Tensor2D, etc.).

[1] https://github.com/tensorflow/tensorflow/issues/31579

File: rfcs/20200211-tf-types.md (hunk 10 above)
Comment (mdanatg) [reply rfcs/20200211-tf-types.md]: Although I haven't tested it, I'd expect Union to work because Tensor is a simple [Generic](https://docs.python.org/3/library/typing.html#typing.Generic) here. If that doesn't work, we could fall back to a less-powerful `AnyDType` identifier.

File: rfcs/20200211-tf-types.md (hunk 11 above)
Comment (fchollet) [reply rfcs/20200211-tf-types.md]: A plain tuple would be much clearer indeed. Perhaps an abstract `rank` as well in case the rank is unknown (e.g. `Shape(rank=None)` would be a dynamic-rank shape.

Comment (fchollet) [reply rfcs/20200211-tf-types.md]: The problem is `Tensor1D`, `Tensor2D` is that it hard-codes the rank, which isn't always a safe assumption. Could there be a type to specify "rank is dynamic but last dimension is N"? (which is what a `Dense` layer would need, for instance).

Comment (mdanatg) [reply rfcs/20200211-tf-types.md]: The inability to mix different ranks is an unfortunate limitation of the type annotation system, as far as I can tell. But specifying an unknown rank or partial shape should be possible.

Comment (gabrieldemarmiesse) [reply rfcs/20200211-tf-types.md]: > The inability to mix different ranks is an unfortunate limitation of the type annotation system, as far as I can tell. 

Could you expand on this? It would be desirable to have something as clear as [typing.Callable](https://docs.python.org/3/library/typing.html#callable). 

If `typing.Callable` can accept
```python
Callable[[str, int], float]
Callable[[None, int, int], Any]
Callable[..., int]
```

What would prevents us from having:

```python
# array of dim 3 or more, dim N°2 is unspecified.
def some_function(x: Tensor[Union[tf.float32, tf.float64], [32, 32, None, ...]]) -> tf.Tensor:
    pass

# array with unknown shape:
def sum(x: Tensor[tf.float32, None], y: Tensor[tf.float32, None]) -> tf.Tensor:
    return x + y

# array with any dtype
def sum(x: Tensor[Any, [32, 32]], y: Tensor[Any, [32, 32]]) -> tf.Tensor:
     return x + y
```

For the `...` we should follow numpy's conventions as much as possible of course.

Comment (mdanatg) [reply rfcs/20200211-tf-types.md]: Interesting, that might work. Poking inside the source code, `Callable` and `Tuple` are special cases of the private `_VariadicGenericAlias`. I think that's what we'd need: https://github.com/python/cpython/blob/3.8/Lib/typing.py#L795

Anyway, the main requirement is contravariance, I think: and value of type `Tensor[Any, [32]]` should be compatible with an argument of type `Tensor[Any, [None]]`. At the same time, `Tensor[Any, [32, 32]]` should not be compatible `Tensor[Any, [None]]` but should be compatible with `Tensor[Any, None]`. How would the type declarations that allow us to to that look like?

Comment (gabrieldemarmiesse) [reply rfcs/20200211-tf-types.md]: >  How would the type declarations that allow us to to that look like?

I'm not sure I understand your question. Are you talking about the API or the implementation details?

Comment (mdanatg) [reply rfcs/20200211-tf-types.md]: I'm asking for a prototype which shows that e.g. `Tensor[tf.float32, [32, 32, None, ...]]` can be implemented.

Comment (gabrieldemarmiesse) [reply rfcs/20200211-tf-types.md]: If we refer to the base implementation of tensorflow/tensorflow#31579 , currently it's not possible due to a limitation of the `input_signature` argument of tf.function which is too strict. For example, it's not possible to set a tf.TensorSpec which can accept multiple datatypes or specify only part of the shape (e.g. it's impossible to specify `[None, ..., 32]` as a valid shape). But there is no real reason why it shouldn't be possible to have a nice API like this. Tensorflow  would be free to redraw the graph if needed like it does when there is no `input_signature` to a tf.function.

Currently, setting an `input_signature` forces a python function to have only a single graph. But it's a big limitation. For example, it's impossible to do:

```python
# this function should make a new graph automatically for each new tensor dtype. 
# So 2 graphs max.
@tf.tfunction(input_signature=(tf.TensorSpec(shape=None, dtype=[tf.float32, tf.float64]),))
def square(x):
    return x**2
```

It's also impossible to specify a python object in a `input_signature` and arbitrary shapes too.

But it's needed badly. We have this issue in tensorflow addons where we can't use tf.function in our API because either it will draw too many graphs, either it will draw a single graph (which is not enough. e.g. multiple possible dtypes for tensors). 

The simplest use case is that you have a public API accepting in the same signature python object (which shouldn't be converted to tensors) and tensors with multiple data types, this happens frequently in tensorflow addons. We just can't use tf.function then.

The bottom line is that we need to change the tf.function API to have better control over the signature (python types, multiple dtype, any dimension for a tensor...), or else, we won't be able to implement and elegant API for type hints with tensorflow.

One possibility would be to go for your implementation proposed in this RFC, and when `input_signature` is improved and gets more flexible, we'll be allowed to use a clearer API and deprecate the old one.

The second possibility is to postpone the implementation of the type hints with tf.function until `input_signature` has more flexibility.

@qlzh727 , if I'm not mistaken, you're familiar with the implementation of tf.function. Do you think improving `input_signature` to allow more flexible signatures is possible?

Comment (mdanatg) [reply rfcs/20200211-tf-types.md]: The RFC definitely doesn't finalize the design for this, so we have time - I'll clarify that in the doc.

I'm not too worried about the input_signature - we can add whatever functionality is missing to align it with type annotations, as long as it remains backward compatible. Backward compatibility isn't hard to achieve I think. The hard limitation is the amount of expressivity that the type annotations allow us, because they're not generic expressions and do have significant limitations. That's why I keep asking whether proposed interfaces are actually realizable in practice.

Comment (gabrieldemarmiesse) [reply rfcs/20200211-tf-types.md]: I'm not familiar enough with the usage of `typing.Generic` to affirm anything concerning the faisability of this :( I think it's possible based on what's available in the standard library, but I'm not positive.

Comment (shoyer) [reply rfcs/20200211-tf-types.md]: With regards to syntax for shape typing, the broader Python ecosystem is very interested in this topic. Hopefully we can figure out a design/syntax that works across a range of array libraries (TensorFlow/NumPy/JAX/PyTorch) and type checkers (mypy/pytype/...).

Here are links to a few related discussions in other projects:
https://github.com/python/typing/issues/516
https://github.com/python/typing/issues/513
https://github.com/numpy/numpy/issues/7370

I would also suggest reaching out to the Python [typing-sig](https://mail.python.org/mailman3/lists/typing-sig.python.org/) mailing list, where this has also been a topic of repeated interest.

Comment (mdanatg) [reply rfcs/20200211-tf-types.md]: Thank for the pointers, yes these are interesting discussions! I think only a few pieces are missing - happy to join the choir.

Comment (bhack) [reply rfcs/20200211-tf-types.md:159]: I will add also https://github.com/google/jax/issues/943
https://github.com/google/jax/pull/2644
@shoyer check also this thread https://llvm.discourse.group/t/tcp-discussion-at-the-mlir-open-meeting-on-4-9/858/
